{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Files for Parsing\n",
    "\n",
    "- Transform json file to dataframe\n",
    "- Transform info file to dataframe\n",
    "- Merge json dataframe with info dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_data_path = \"/Users/claudia/Downloads/dataset2.json\"\n",
    "# info_data_path = \"/Users/claudia/Downloads/data.info\"\n",
    "data = [json.loads(line) for line in open(json_data_path, 'r')]\n",
    "# with open(info_data_path, 'r') as f:\n",
    "#     info = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## function to get key of a dictionary\n",
    "def get_key(dictionary):\n",
    "    key_object = dictionary.keys()\n",
    "    key = list(key_object)[0]\n",
    "    return key\n",
    "\n",
    "## function to help concatenate columns to get transcript, position, nucleotides\n",
    "def concat_col(transcript, position, nucleotide, n):\n",
    "    t_df = pd.DataFrame([transcript]*n)\n",
    "    p_df = pd.DataFrame([position]*n)\n",
    "    nu_df = pd.DataFrame([nucleotide]*n)\n",
    "    n_df = pd.DataFrame([n]*n)\n",
    "\n",
    "    ## concat columns together\n",
    "    final_df = pd.concat([t_df, p_df, nu_df, n_df], axis = 1)\n",
    "    final_df.columns = ['transcript', 'position', 'nucleotides', 'reads_count']\n",
    "    return final_df\n",
    "\n",
    "## function to parse line in json file\n",
    "def parse_line(line):\n",
    "    ## get transcript\n",
    "    t = get_key(line)\n",
    "\n",
    "    ## get position\n",
    "    p = get_key(line[t])\n",
    "\n",
    "    ## get nucleotide seq\n",
    "    nu = get_key(line[t][p])\n",
    "\n",
    "    ## get number of reads\n",
    "    reads_count = len(line[t][p][nu])\n",
    "\n",
    "    ## get dataframe of list of reads\n",
    "    reads = pd.DataFrame(line[t][p][nu])\n",
    "\n",
    "    ## concat columns together to get transcript, position, nucleotides and all dwelling time, std, mean\n",
    "    df = pd.concat([concat_col(t, p, nu, reads_count), reads], axis = 1)\n",
    "    df.columns = ['transcript', 'position', 'nucleotides', 'reads_count', 'dwellingtime_-1', 'std_-1', 'mean_-1', 'dwellingtime_0', 'std_0', 'mean_0', 'dwellingtime_+1', 'std_+1', 'mean_+1']\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>dwellingtime_-1</th>\n",
       "      <th>std_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwellingtime_0</th>\n",
       "      <th>std_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwellingtime_+1</th>\n",
       "      <th>std_+1</th>\n",
       "      <th>mean_+1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>1.93</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01260</td>\n",
       "      <td>1.97</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00421</td>\n",
       "      <td>1.50</td>\n",
       "      <td>95.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>2.27</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00536</td>\n",
       "      <td>2.49</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>2.28</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>6.29</td>\n",
       "      <td>114.0</td>\n",
       "      <td>0.00442</td>\n",
       "      <td>2.07</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00785</td>\n",
       "      <td>1.97</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00617</td>\n",
       "      <td>5.16</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>2.70</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>2.82</td>\n",
       "      <td>97.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.01</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00495</td>\n",
       "      <td>1.89</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.01100</td>\n",
       "      <td>1.64</td>\n",
       "      <td>97.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>1.94</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00520</td>\n",
       "      <td>3.53</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00365</td>\n",
       "      <td>1.73</td>\n",
       "      <td>96.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>2.34</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00673</td>\n",
       "      <td>2.73</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00570</td>\n",
       "      <td>1.93</td>\n",
       "      <td>95.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.42</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>1.65</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00547</td>\n",
       "      <td>1.53</td>\n",
       "      <td>95.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00724</td>\n",
       "      <td>2.40</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>2.98</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00474</td>\n",
       "      <td>1.45</td>\n",
       "      <td>92.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.01930</td>\n",
       "      <td>1.68</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00758</td>\n",
       "      <td>2.73</td>\n",
       "      <td>113.0</td>\n",
       "      <td>0.00524</td>\n",
       "      <td>2.13</td>\n",
       "      <td>93.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>1.26</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00788</td>\n",
       "      <td>2.36</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00463</td>\n",
       "      <td>1.69</td>\n",
       "      <td>95.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>1.86</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00777</td>\n",
       "      <td>2.64</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00581</td>\n",
       "      <td>1.55</td>\n",
       "      <td>93.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.01330</td>\n",
       "      <td>2.54</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.01130</td>\n",
       "      <td>1.99</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00467</td>\n",
       "      <td>1.53</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>1.34</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00591</td>\n",
       "      <td>2.09</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>1.72</td>\n",
       "      <td>96.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>1.83</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01140</td>\n",
       "      <td>2.28</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00697</td>\n",
       "      <td>2.35</td>\n",
       "      <td>95.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00963</td>\n",
       "      <td>2.08</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00283</td>\n",
       "      <td>1.88</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>1.88</td>\n",
       "      <td>97.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>1.28</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.59</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.00659</td>\n",
       "      <td>1.38</td>\n",
       "      <td>96.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.10</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00863</td>\n",
       "      <td>2.74</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>2.28</td>\n",
       "      <td>95.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.01590</td>\n",
       "      <td>2.01</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>1.99</td>\n",
       "      <td>99.5</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>1.47</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>1.84</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00920</td>\n",
       "      <td>2.64</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.00910</td>\n",
       "      <td>1.98</td>\n",
       "      <td>94.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>6.91</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00494</td>\n",
       "      <td>1.55</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00484</td>\n",
       "      <td>2.60</td>\n",
       "      <td>93.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.85</td>\n",
       "      <td>113.0</td>\n",
       "      <td>0.00679</td>\n",
       "      <td>2.86</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>1.68</td>\n",
       "      <td>95.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>3.43</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.01060</td>\n",
       "      <td>3.30</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>2.21</td>\n",
       "      <td>98.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>1.25</td>\n",
       "      <td>113.0</td>\n",
       "      <td>0.00599</td>\n",
       "      <td>1.59</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00786</td>\n",
       "      <td>2.00</td>\n",
       "      <td>98.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.67</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01330</td>\n",
       "      <td>2.01</td>\n",
       "      <td>98.3</td>\n",
       "      <td>0.00431</td>\n",
       "      <td>1.53</td>\n",
       "      <td>94.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>1.39</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00549</td>\n",
       "      <td>2.21</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01140</td>\n",
       "      <td>1.60</td>\n",
       "      <td>97.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>1.11</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00406</td>\n",
       "      <td>2.00</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00963</td>\n",
       "      <td>1.99</td>\n",
       "      <td>94.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>1.00</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.00860</td>\n",
       "      <td>2.00</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.00738</td>\n",
       "      <td>1.57</td>\n",
       "      <td>97.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>1.66</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00491</td>\n",
       "      <td>2.31</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>1.33</td>\n",
       "      <td>94.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>2.30</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00688</td>\n",
       "      <td>2.04</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00333</td>\n",
       "      <td>1.76</td>\n",
       "      <td>94.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>3.45</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.01060</td>\n",
       "      <td>2.60</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00747</td>\n",
       "      <td>1.87</td>\n",
       "      <td>94.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>1.90</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>2.15</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.00291</td>\n",
       "      <td>1.26</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.02130</td>\n",
       "      <td>4.03</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>1.57</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.00675</td>\n",
       "      <td>1.91</td>\n",
       "      <td>94.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00365</td>\n",
       "      <td>4.51</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.00479</td>\n",
       "      <td>1.71</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.01860</td>\n",
       "      <td>1.63</td>\n",
       "      <td>97.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.01070</td>\n",
       "      <td>2.30</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.01460</td>\n",
       "      <td>2.11</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00843</td>\n",
       "      <td>1.72</td>\n",
       "      <td>97.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00365</td>\n",
       "      <td>3.73</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.01650</td>\n",
       "      <td>2.08</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00388</td>\n",
       "      <td>1.48</td>\n",
       "      <td>93.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript position nucleotides  reads_count  dwellingtime_-1  std_-1  \\\n",
       "0   AT1G01050.1      155     GAAACTA           36          0.00232    1.93   \n",
       "1   AT1G01050.1      155     GAAACTA           36          0.00896    2.27   \n",
       "2   AT1G01050.1      155     GAAACTA           36          0.00498    6.29   \n",
       "3   AT1G01050.1      155     GAAACTA           36          0.00617    5.16   \n",
       "4   AT1G01050.1      155     GAAACTA           36          0.00664    2.01   \n",
       "5   AT1G01050.1      155     GAAACTA           36          0.00631    1.94   \n",
       "6   AT1G01050.1      155     GAAACTA           36          0.00232    2.34   \n",
       "7   AT1G01050.1      155     GAAACTA           36          0.00498    2.42   \n",
       "8   AT1G01050.1      155     GAAACTA           36          0.00724    2.40   \n",
       "9   AT1G01050.1      155     GAAACTA           36          0.01930    1.68   \n",
       "10  AT1G01050.1      155     GAAACTA           36          0.00863    1.26   \n",
       "11  AT1G01050.1      155     GAAACTA           36          0.00531    1.86   \n",
       "12  AT1G01050.1      155     GAAACTA           36          0.01330    2.54   \n",
       "13  AT1G01050.1      155     GAAACTA           36          0.00232    1.34   \n",
       "14  AT1G01050.1      155     GAAACTA           36          0.00432    1.83   \n",
       "15  AT1G01050.1      155     GAAACTA           36          0.00963    2.08   \n",
       "16  AT1G01050.1      155     GAAACTA           36          0.00299    1.28   \n",
       "17  AT1G01050.1      155     GAAACTA           36          0.00664    2.10   \n",
       "18  AT1G01050.1      155     GAAACTA           36          0.01590    2.01   \n",
       "19  AT1G01050.1      155     GAAACTA           36          0.00266    1.84   \n",
       "20  AT1G01050.1      155     GAAACTA           36          0.00398    6.91   \n",
       "21  AT1G01050.1      155     GAAACTA           36          0.00266    3.85   \n",
       "22  AT1G01050.1      155     GAAACTA           36          0.00432    3.43   \n",
       "23  AT1G01050.1      155     GAAACTA           36          0.00432    1.25   \n",
       "24  AT1G01050.1      155     GAAACTA           36          0.00631    2.67   \n",
       "25  AT1G01050.1      155     GAAACTA           36          0.00564    1.39   \n",
       "26  AT1G01050.1      155     GAAACTA           36          0.00531    1.11   \n",
       "27  AT1G01050.1      155     GAAACTA           36          0.00465    1.00   \n",
       "28  AT1G01050.1      155     GAAACTA           36          0.00531    1.66   \n",
       "29  AT1G01050.1      155     GAAACTA           36          0.00531    2.30   \n",
       "30  AT1G01050.1      155     GAAACTA           36          0.00432    3.45   \n",
       "31  AT1G01050.1      155     GAAACTA           36          0.00432    1.90   \n",
       "32  AT1G01050.1      155     GAAACTA           36          0.02130    4.03   \n",
       "33  AT1G01050.1      155     GAAACTA           36          0.00365    4.51   \n",
       "34  AT1G01050.1      155     GAAACTA           36          0.01070    2.30   \n",
       "35  AT1G01050.1      155     GAAACTA           36          0.00365    3.73   \n",
       "\n",
       "    mean_-1  dwellingtime_0  std_0  mean_0  dwellingtime_+1  std_+1  mean_+1  \n",
       "0     109.0         0.01260   1.97   111.0          0.00421    1.50     95.3  \n",
       "1     110.0         0.00536   2.49   110.0          0.00797    2.28     96.0  \n",
       "2     114.0         0.00442   2.07   111.0          0.00785    1.97     96.0  \n",
       "3     106.0         0.00830   2.70   105.0          0.00199    2.82     97.4  \n",
       "4     110.0         0.00495   1.89   110.0          0.01100    1.64     97.3  \n",
       "5     110.0         0.00520   3.53   109.0          0.00365    1.73     96.2  \n",
       "6     106.0         0.00673   2.73   109.0          0.00570    1.93     95.3  \n",
       "7     110.0         0.00199   1.65   102.0          0.00547    1.53     95.2  \n",
       "8     109.0         0.00531   2.98   106.0          0.00474    1.45     92.7  \n",
       "9     110.0         0.00758   2.73   113.0          0.00524    2.13     93.7  \n",
       "10    110.0         0.00788   2.36   109.0          0.00463    1.69     95.7  \n",
       "11    110.0         0.00777   2.64   111.0          0.00581    1.55     93.3  \n",
       "12    111.0         0.01130   1.99   110.0          0.00467    1.53     95.0  \n",
       "13    111.0         0.00591   2.09   111.0          0.00465    1.72     96.1  \n",
       "14    109.0         0.01140   2.28   110.0          0.00697    2.35     95.9  \n",
       "15    110.0         0.00283   1.88   110.0          0.00863    1.88     97.6  \n",
       "16    108.0         0.00398   2.59   112.0          0.00659    1.38     96.3  \n",
       "17    110.0         0.00863   2.74   110.0          0.00797    2.28     95.4  \n",
       "18    111.0         0.00531   1.99    99.5          0.00332    1.47     95.0  \n",
       "19    108.0         0.00920   2.64   107.0          0.00910    1.98     94.9  \n",
       "20    109.0         0.00494   1.55   111.0          0.00484    2.60     93.6  \n",
       "21    113.0         0.00679   2.86   108.0          0.00531    1.68     95.6  \n",
       "22    111.0         0.01060   3.30   111.0          0.00266    2.21     98.2  \n",
       "23    113.0         0.00599   1.59   108.0          0.00786    2.00     98.6  \n",
       "24    109.0         0.01330   2.01    98.3          0.00431    1.53     94.7  \n",
       "25    111.0         0.00549   2.21   109.0          0.01140    1.60     97.1  \n",
       "26    110.0         0.00406   2.00   110.0          0.00963    1.99     94.6  \n",
       "27    112.0         0.00860   2.00   108.0          0.00738    1.57     97.1  \n",
       "28    106.0         0.00491   2.31   110.0          0.00332    1.33     94.1  \n",
       "29    110.0         0.00688   2.04   110.0          0.00333    1.76     94.7  \n",
       "30    110.0         0.01060   2.60   109.0          0.00747    1.87     94.7  \n",
       "31    109.0         0.00316   2.15   107.0          0.00291    1.26     96.0  \n",
       "32    112.0         0.00266   1.57   107.0          0.00675    1.91     94.5  \n",
       "33    109.0         0.00479   1.71   112.0          0.01860    1.63     97.6  \n",
       "34    112.0         0.01460   2.11   110.0          0.00843    1.72     97.9  \n",
       "35    108.0         0.01650   2.08   110.0          0.00388    1.48     93.8  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_line(data[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## parse all lines in json file into dataframe for concatenation\n",
    "reads_df = [parse_line(data[i]) for i in range(len(data))]\n",
    "data_df = pd.concat(reads_df, axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in data_df = 6903936\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>dwellingtime_-1</th>\n",
       "      <th>std_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwellingtime_0</th>\n",
       "      <th>std_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwellingtime_+1</th>\n",
       "      <th>std_+1</th>\n",
       "      <th>mean_+1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>1.93</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01260</td>\n",
       "      <td>1.97</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00421</td>\n",
       "      <td>1.50</td>\n",
       "      <td>95.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>2.27</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00536</td>\n",
       "      <td>2.49</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>2.28</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>6.29</td>\n",
       "      <td>114.0</td>\n",
       "      <td>0.00442</td>\n",
       "      <td>2.07</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00785</td>\n",
       "      <td>1.97</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00617</td>\n",
       "      <td>5.16</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>2.70</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>2.82</td>\n",
       "      <td>97.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.01</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00495</td>\n",
       "      <td>1.89</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.01100</td>\n",
       "      <td>1.64</td>\n",
       "      <td>97.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript position nucleotides  reads_count  dwellingtime_-1  std_-1  \\\n",
       "0  AT1G01050.1      155     GAAACTA           36          0.00232    1.93   \n",
       "1  AT1G01050.1      155     GAAACTA           36          0.00896    2.27   \n",
       "2  AT1G01050.1      155     GAAACTA           36          0.00498    6.29   \n",
       "3  AT1G01050.1      155     GAAACTA           36          0.00617    5.16   \n",
       "4  AT1G01050.1      155     GAAACTA           36          0.00664    2.01   \n",
       "\n",
       "   mean_-1  dwellingtime_0  std_0  mean_0  dwellingtime_+1  std_+1  mean_+1  \n",
       "0    109.0         0.01260   1.97   111.0          0.00421    1.50     95.3  \n",
       "1    110.0         0.00536   2.49   110.0          0.00797    2.28     96.0  \n",
       "2    114.0         0.00442   2.07   111.0          0.00785    1.97     96.0  \n",
       "3    106.0         0.00830   2.70   105.0          0.00199    2.82     97.4  \n",
       "4    110.0         0.00495   1.89   110.0          0.01100    1.64     97.3  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Number of rows in data_df = {len(data_df)}\")\n",
    "data_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id transcript_position label\n",
       "0  ENSG00000004059  ENST00000000233                 244     0\n",
       "1  ENSG00000004059  ENST00000000233                 261     0\n",
       "2  ENSG00000004059  ENST00000000233                 316     0\n",
       "3  ENSG00000004059  ENST00000000233                 332     0\n",
       "4  ENSG00000004059  ENST00000000233                 368     0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## transform info data into dataframe\n",
    "info_list = [info[i].split(\",\") for i in range(len(info))]\n",
    "info_df = pd.DataFrame(info_list[1:]) \n",
    "info_df.columns = info_list[0]\n",
    "info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>dwellingtime_-1</th>\n",
       "      <th>std_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwellingtime_0</th>\n",
       "      <th>std_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwellingtime_+1</th>\n",
       "      <th>std_+1</th>\n",
       "      <th>mean_+1</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript position nucleotides  reads_count  dwellingtime_-1  std_-1  \\\n",
       "0  ENST00000000233      244     AAGACCA          185          0.00299    2.06   \n",
       "1  ENST00000000233      244     AAGACCA          185          0.00631    2.53   \n",
       "2  ENST00000000233      244     AAGACCA          185          0.00465    3.92   \n",
       "3  ENST00000000233      244     AAGACCA          185          0.00398    2.06   \n",
       "4  ENST00000000233      244     AAGACCA          185          0.00664    2.92   \n",
       "\n",
       "   mean_-1  dwellingtime_0  std_0  mean_0  dwellingtime_+1  std_+1  mean_+1  \\\n",
       "0    125.0         0.01770  10.40   122.0          0.00930   10.90     84.1   \n",
       "1    125.0         0.00844   4.67   126.0          0.01030    6.30     80.9   \n",
       "2    109.0         0.01360  12.00   124.0          0.00498    2.13     79.6   \n",
       "3    125.0         0.00830   5.01   130.0          0.00498    3.78     80.4   \n",
       "4    120.0         0.00266   3.94   129.0          0.01300    7.15     82.2   \n",
       "\n",
       "           gene_id    transcript_id transcript_position label  \n",
       "0  ENSG00000004059  ENST00000000233                 244     0  \n",
       "1  ENSG00000004059  ENST00000000233                 244     0  \n",
       "2  ENSG00000004059  ENST00000000233                 244     0  \n",
       "3  ENSG00000004059  ENST00000000233                 244     0  \n",
       "4  ENSG00000004059  ENST00000000233                 244     0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## left join data_df and info_df to get features and labels\n",
    "complete_df = data_df.merge(info_df, how = \"left\", left_on = [\"transcript\", \"position\"], right_on = [\"transcript_id\", \"transcript_position\"])\n",
    "complete_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>dwellingtime_-1</th>\n",
       "      <th>std_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwellingtime_0</th>\n",
       "      <th>std_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwellingtime_+1</th>\n",
       "      <th>std_+1</th>\n",
       "      <th>mean_+1</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>185</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript position nucleotides  reads_count  dwellingtime_-1  std_-1  \\\n",
       "0  ENST00000000233      244     AAGACCA          185          0.00299    2.06   \n",
       "1  ENST00000000233      244     AAGACCA          185          0.00631    2.53   \n",
       "2  ENST00000000233      244     AAGACCA          185          0.00465    3.92   \n",
       "3  ENST00000000233      244     AAGACCA          185          0.00398    2.06   \n",
       "4  ENST00000000233      244     AAGACCA          185          0.00664    2.92   \n",
       "\n",
       "   mean_-1  dwellingtime_0  std_0  mean_0  dwellingtime_+1  std_+1  mean_+1  \\\n",
       "0    125.0         0.01770  10.40   122.0          0.00930   10.90     84.1   \n",
       "1    125.0         0.00844   4.67   126.0          0.01030    6.30     80.9   \n",
       "2    109.0         0.01360  12.00   124.0          0.00498    2.13     79.6   \n",
       "3    125.0         0.00830   5.01   130.0          0.00498    3.78     80.4   \n",
       "4    120.0         0.00266   3.94   129.0          0.01300    7.15     82.2   \n",
       "\n",
       "           gene_id label  \n",
       "0  ENSG00000004059     0  \n",
       "1  ENSG00000004059     0  \n",
       "2  ENSG00000004059     0  \n",
       "3  ENSG00000004059     0  \n",
       "4  ENSG00000004059     0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## drop duplicate transcript_id and transcript_position column\n",
    "complete_df.drop(['transcript_id', 'transcript_position'],  axis = 1, inplace = True)\n",
    "complete_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving complete_df into parquet files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## function to breakdown dataframe into smaller sizes and save it\n",
    "def save_file(df, filename, nrows = 5000000):\n",
    "    total_rows = len(df)\n",
    "    start, stop, count = 0, nrows, 1\n",
    "\n",
    "    ## use while loop to break the dataframe into smaller dataframes\n",
    "    while stop < total_rows:\n",
    "        print(start, stop)\n",
    "        temp_df = pd.DataFrame(df.iloc[start:stop, :])\n",
    "        fname = f\"{filename}_{count}.parquet\"\n",
    "        temp_df.to_parquet(fname)\n",
    "        print(f\"Saved a file called {fname}\")\n",
    "        count += 1\n",
    "        start += nrows\n",
    "        stop += nrows\n",
    "    \n",
    "    stop = total_rows\n",
    "    print(start, stop)\n",
    "    temp_df = pd.DataFrame(df.iloc[start:stop, :])\n",
    "    fname = f\"{filename}_{count}.parquet\"\n",
    "    temp_df.to_parquet(fname)\n",
    "    print(f\"Saved a file called {fname}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 5000000\n",
      "Saved a file called dataset2_1.parquet\n",
      "5000000 6903936\n",
      "Saved a file called dataset2_2.parquet\n"
     ]
    }
   ],
   "source": [
    "save_file(data_df, \"dataset2\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.8 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
