{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Importing Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\lowmi\\OneDrive\\Desktop\\Uni\\DSA4262\\DSA4262-ACMXZ\n"
     ]
    }
   ],
   "source": [
    "%cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from util.preprocess import (Preprocessing)\n",
    "from util.preprocess.parameters import model_features_list\n",
    "from util.model.training import get_percent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_1 = [json.loads(line) for line in open(\"data/raw_data/dataset1.json\", 'r')]\n",
    "data_2 = [json.loads(line) for line in open(\"data/raw_data/dataset2.json\", 'r')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/raw_data/data.info\", 'r') as f:\n",
    "    info = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# encoding pipeline\n",
    "infile = open(\"data/raw_data/encoding_pipeline.pkl\",'rb')\n",
    "enc_pipe = pickle.load(infile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "## function to get key of a dictionary\n",
    "def get_key(dictionary):\n",
    "    key_object = dictionary.keys()\n",
    "    key = list(key_object)[0]\n",
    "    return key\n",
    "\n",
    "## function to help concatenate columns to get transcript, position, nucleotides\n",
    "def concat_col(transcript, position, nucleotide, n):\n",
    "    t_df = pd.DataFrame([transcript]*n)\n",
    "    p_df = pd.DataFrame([position]*n)\n",
    "    nu_df = pd.DataFrame([nucleotide]*n)\n",
    "    n_df = pd.DataFrame([n]*n)\n",
    "\n",
    "    ## concat columns together\n",
    "    final_df = pd.concat([t_df, p_df, nu_df, n_df], axis = 1)\n",
    "    final_df.columns = ['transcript', 'position', 'nucleotides', 'reads_count']\n",
    "    return final_df\n",
    "\n",
    "## function to parse line in json file\n",
    "def parse_line(line):\n",
    "    ## get transcript\n",
    "    t = get_key(line)\n",
    "\n",
    "    ## get position\n",
    "    p = get_key(line[t])\n",
    "\n",
    "    ## get nucleotide seq\n",
    "    nu = get_key(line[t][p])\n",
    "\n",
    "    ## get number of reads\n",
    "    reads_count = len(line[t][p][nu])\n",
    "\n",
    "    ## get dataframe of list of reads\n",
    "    reads = pd.DataFrame(line[t][p][nu])\n",
    "\n",
    "    ## concat columns together to get transcript, position, nucleotides and all dwelling time, std, mean\n",
    "    df = pd.concat([concat_col(t, p, nu, reads_count), reads], axis = 1)\n",
    "    df.columns = ['transcript', 'position', 'nucleotides', 'reads_count', 'dwellingtime_-1', 'std_-1', 'mean_-1', 'dwellingtime_0', 'std_0', 'mean_0', 'dwellingtime_+1', 'std_+1', 'mean_+1']\n",
    "\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse all lines in data_1 file into dataframe for concatenation\n",
    "reads_1 = [parse_line(data_1[i]) for i in range(len(data_1))]\n",
    "df_1 = pd.concat(reads_1, axis = 0)\n",
    "\n",
    "# parse all lines in data_2 file into dataframe for concatenation\n",
    "reads_2 = [parse_line(data_2[i]) for i in range(len(data_2))]\n",
    "df_2 = pd.concat(reads_2, axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform info data into dataframe\n",
    "info_list = [info[i].split(\",\") for i in range(len(info))]\n",
    "info_df = pd.DataFrame(info_list[1:]) \n",
    "info_df.columns = info_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# left join data_df and info_df to get features and labels\n",
    "df_1 = df_1.merge(info_df, how = \"left\", left_on = [\"transcript\", \"position\"], right_on = [\"transcript_id\", \"transcript_position\"])\n",
    "df_2 = df_2.merge(info_df, how = \"left\", left_on = [\"transcript\", \"position\"], right_on = [\"transcript_id\", \"transcript_position\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicate transcript_id and transcript_position column\n",
    "df_1.drop(['transcript_id', 'transcript_position'],  axis = 1, inplace = True)\n",
    "df_2.drop(['transcript_id', 'transcript_position'],  axis = 1, inplace = True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_1_pre = Preprocessing(df_1)\n",
    "df_2_pre = Preprocessing(df_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>dwellingtime_-1</th>\n",
       "      <th>std_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwellingtime_0</th>\n",
       "      <th>std_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwellingtime_+1</th>\n",
       "      <th>std_+1</th>\n",
       "      <th>mean_+1</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>1.93</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01260</td>\n",
       "      <td>1.97</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00421</td>\n",
       "      <td>1.50</td>\n",
       "      <td>95.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>2.27</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00536</td>\n",
       "      <td>2.49</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>2.28</td>\n",
       "      <td>96.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>6.29</td>\n",
       "      <td>114.0</td>\n",
       "      <td>0.00442</td>\n",
       "      <td>2.07</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00785</td>\n",
       "      <td>1.97</td>\n",
       "      <td>96.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00617</td>\n",
       "      <td>5.16</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>2.70</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>2.82</td>\n",
       "      <td>97.4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.01</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00495</td>\n",
       "      <td>1.89</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.01100</td>\n",
       "      <td>1.64</td>\n",
       "      <td>97.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903931</th>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01080</td>\n",
       "      <td>7.15</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.00410</td>\n",
       "      <td>1.46</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>1.46</td>\n",
       "      <td>102.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903932</th>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02460</td>\n",
       "      <td>5.91</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.00836</td>\n",
       "      <td>1.77</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>3.19</td>\n",
       "      <td>88.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903933</th>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00842</td>\n",
       "      <td>8.67</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.00697</td>\n",
       "      <td>5.23</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.00978</td>\n",
       "      <td>1.62</td>\n",
       "      <td>95.6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903934</th>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>6.28</td>\n",
       "      <td>135.0</td>\n",
       "      <td>0.00712</td>\n",
       "      <td>1.91</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>3.04</td>\n",
       "      <td>96.8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903935</th>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00526</td>\n",
       "      <td>5.91</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>3.72</td>\n",
       "      <td>106.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>1.67</td>\n",
       "      <td>99.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6903936 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript position nucleotides  reads_count gene_id  \\\n",
       "0        AT1G01050.1      155     GAAACTA           36     NaN   \n",
       "1        AT1G01050.1      155     GAAACTA           36     NaN   \n",
       "2        AT1G01050.1      155     GAAACTA           36     NaN   \n",
       "3        AT1G01050.1      155     GAAACTA           36     NaN   \n",
       "4        AT1G01050.1      155     GAAACTA           36     NaN   \n",
       "...              ...      ...         ...          ...     ...   \n",
       "6903931  AT5G67600.1      154     AGAACTC           30     NaN   \n",
       "6903932  AT5G67600.1      154     AGAACTC           30     NaN   \n",
       "6903933  AT5G67600.1      154     AGAACTC           30     NaN   \n",
       "6903934  AT5G67600.1      154     AGAACTC           30     NaN   \n",
       "6903935  AT5G67600.1      154     AGAACTC           30     NaN   \n",
       "\n",
       "         dwellingtime_-1  std_-1  mean_-1  dwellingtime_0  std_0  mean_0  \\\n",
       "0                0.00232    1.93    109.0         0.01260   1.97   111.0   \n",
       "1                0.00896    2.27    110.0         0.00536   2.49   110.0   \n",
       "2                0.00498    6.29    114.0         0.00442   2.07   111.0   \n",
       "3                0.00617    5.16    106.0         0.00830   2.70   105.0   \n",
       "4                0.00664    2.01    110.0         0.00495   1.89   110.0   \n",
       "...                  ...     ...      ...             ...    ...     ...   \n",
       "6903931          0.01080    7.15    136.0         0.00410   1.46   106.0   \n",
       "6903932          0.02460    5.91    136.0         0.00836   1.77   105.0   \n",
       "6903933          0.00842    8.67    136.0         0.00697   5.23   110.0   \n",
       "6903934          0.00316    6.28    135.0         0.00712   1.91   105.0   \n",
       "6903935          0.00526    5.91    136.0         0.00299   3.72   106.0   \n",
       "\n",
       "         dwellingtime_+1  std_+1  mean_+1  label  \n",
       "0                0.00421    1.50     95.3      2  \n",
       "1                0.00797    2.28     96.0      2  \n",
       "2                0.00785    1.97     96.0      2  \n",
       "3                0.00199    2.82     97.4      2  \n",
       "4                0.01100    1.64     97.3      2  \n",
       "...                  ...     ...      ...    ...  \n",
       "6903931          0.00930    1.46    102.0      2  \n",
       "6903932          0.00664    3.19     88.3      2  \n",
       "6903933          0.00978    1.62     95.6      2  \n",
       "6903934          0.00232    3.04     96.8      2  \n",
       "6903935          0.00266    1.67     99.0      2  \n",
       "\n",
       "[6903936 rows x 15 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1_pre.model_features_and_clean()\n",
    "df_2_pre.model_features_and_clean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_2_pre.df['gene_id'] = df_2_pre.df['gene_id'].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>label</th>\n",
       "      <th>dwelling_time_-1_25</th>\n",
       "      <th>dwelling_time_-1_50</th>\n",
       "      <th>dwelling_time_-1_75</th>\n",
       "      <th>dwelling_time_-1_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>dwelling_time_+1_75</th>\n",
       "      <th>dwelling_time_+1_mean</th>\n",
       "      <th>std_+1_25</th>\n",
       "      <th>std_+1_50</th>\n",
       "      <th>std_+1_75</th>\n",
       "      <th>std_+1_mean</th>\n",
       "      <th>mean_+1_25</th>\n",
       "      <th>mean_+1_50</th>\n",
       "      <th>mean_+1_75</th>\n",
       "      <th>mean_+1_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004235</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.006790</td>\n",
       "      <td>0.006574</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007888</td>\n",
       "      <td>0.006340</td>\n",
       "      <td>1.5300</td>\n",
       "      <td>1.720</td>\n",
       "      <td>1.9825</td>\n",
       "      <td>1.804167</td>\n",
       "      <td>94.700</td>\n",
       "      <td>95.5</td>\n",
       "      <td>96.500</td>\n",
       "      <td>95.641667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>165</td>\n",
       "      <td>ATAACCA</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004980</td>\n",
       "      <td>0.005705</td>\n",
       "      <td>0.008533</td>\n",
       "      <td>0.006866</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007065</td>\n",
       "      <td>0.005963</td>\n",
       "      <td>1.6100</td>\n",
       "      <td>2.105</td>\n",
       "      <td>2.5400</td>\n",
       "      <td>2.255000</td>\n",
       "      <td>81.800</td>\n",
       "      <td>82.2</td>\n",
       "      <td>83.375</td>\n",
       "      <td>82.560526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>347</td>\n",
       "      <td>AAGACAG</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.003835</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.010280</td>\n",
       "      <td>0.008893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009300</td>\n",
       "      <td>0.008132</td>\n",
       "      <td>2.4700</td>\n",
       "      <td>3.610</td>\n",
       "      <td>5.0800</td>\n",
       "      <td>4.088857</td>\n",
       "      <td>76.600</td>\n",
       "      <td>77.8</td>\n",
       "      <td>79.300</td>\n",
       "      <td>78.120000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>435</td>\n",
       "      <td>AAGACAA</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.006970</td>\n",
       "      <td>0.008465</td>\n",
       "      <td>0.007647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009065</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>2.5100</td>\n",
       "      <td>3.460</td>\n",
       "      <td>4.6950</td>\n",
       "      <td>3.717714</td>\n",
       "      <td>79.900</td>\n",
       "      <td>80.9</td>\n",
       "      <td>84.050</td>\n",
       "      <td>81.951429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>549</td>\n",
       "      <td>ATGACAA</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004105</td>\n",
       "      <td>0.007120</td>\n",
       "      <td>0.011450</td>\n",
       "      <td>0.008467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008030</td>\n",
       "      <td>0.008201</td>\n",
       "      <td>1.8800</td>\n",
       "      <td>2.740</td>\n",
       "      <td>3.8750</td>\n",
       "      <td>2.998000</td>\n",
       "      <td>85.000</td>\n",
       "      <td>88.2</td>\n",
       "      <td>89.350</td>\n",
       "      <td>86.457143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98825</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>367</td>\n",
       "      <td>AAAACTC</td>\n",
       "      <td>117</td>\n",
       "      <td>2</td>\n",
       "      <td>0.007160</td>\n",
       "      <td>0.010200</td>\n",
       "      <td>0.014500</td>\n",
       "      <td>0.011100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008630</td>\n",
       "      <td>0.006633</td>\n",
       "      <td>2.0400</td>\n",
       "      <td>2.330</td>\n",
       "      <td>2.6900</td>\n",
       "      <td>2.416410</td>\n",
       "      <td>94.300</td>\n",
       "      <td>95.4</td>\n",
       "      <td>96.400</td>\n",
       "      <td>95.182051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98826</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>444</td>\n",
       "      <td>GTGACCC</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004430</td>\n",
       "      <td>0.006190</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006270</td>\n",
       "      <td>0.005211</td>\n",
       "      <td>2.3300</td>\n",
       "      <td>3.090</td>\n",
       "      <td>4.1400</td>\n",
       "      <td>3.472566</td>\n",
       "      <td>78.300</td>\n",
       "      <td>80.3</td>\n",
       "      <td>82.100</td>\n",
       "      <td>80.369912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98827</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>465</td>\n",
       "      <td>GTGACTC</td>\n",
       "      <td>118</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004320</td>\n",
       "      <td>0.006340</td>\n",
       "      <td>0.009132</td>\n",
       "      <td>0.007294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.006698</td>\n",
       "      <td>2.0350</td>\n",
       "      <td>2.885</td>\n",
       "      <td>3.8300</td>\n",
       "      <td>3.127305</td>\n",
       "      <td>88.025</td>\n",
       "      <td>89.2</td>\n",
       "      <td>90.300</td>\n",
       "      <td>88.889831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98828</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>663</td>\n",
       "      <td>GAAACCT</td>\n",
       "      <td>101</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004980</td>\n",
       "      <td>0.007060</td>\n",
       "      <td>0.010800</td>\n",
       "      <td>0.008229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.006016</td>\n",
       "      <td>2.0400</td>\n",
       "      <td>2.380</td>\n",
       "      <td>2.7700</td>\n",
       "      <td>2.483366</td>\n",
       "      <td>81.200</td>\n",
       "      <td>82.5</td>\n",
       "      <td>83.900</td>\n",
       "      <td>82.502970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98829</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004985</td>\n",
       "      <td>0.007360</td>\n",
       "      <td>0.009405</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007857</td>\n",
       "      <td>0.005726</td>\n",
       "      <td>1.5375</td>\n",
       "      <td>1.765</td>\n",
       "      <td>2.6825</td>\n",
       "      <td>2.047667</td>\n",
       "      <td>93.700</td>\n",
       "      <td>94.5</td>\n",
       "      <td>95.550</td>\n",
       "      <td>94.836667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98830 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id   transcript position nucleotides  reads_count  label  \\\n",
       "0            0  AT1G01050.1      155     GAAACTA           36      2   \n",
       "1            0  AT1G01050.1      165     ATAACCA           38      2   \n",
       "2            0  AT1G01050.1      347     AAGACAG           35      2   \n",
       "3            0  AT1G01050.1      435     AAGACAA           35      2   \n",
       "4            0  AT1G01050.1      549     ATGACAA           35      2   \n",
       "...        ...          ...      ...         ...          ...    ...   \n",
       "98825        0  AT5G67590.1      367     AAAACTC          117      2   \n",
       "98826        0  AT5G67590.1      444     GTGACCC          113      2   \n",
       "98827        0  AT5G67590.1      465     GTGACTC          118      2   \n",
       "98828        0  AT5G67590.1      663     GAAACCT          101      2   \n",
       "98829        0  AT5G67600.1      154     AGAACTC           30      2   \n",
       "\n",
       "       dwelling_time_-1_25  dwelling_time_-1_50  dwelling_time_-1_75  \\\n",
       "0                 0.004235             0.005310             0.006790   \n",
       "1                 0.004980             0.005705             0.008533   \n",
       "2                 0.003835             0.007300             0.010280   \n",
       "3                 0.005310             0.006970             0.008465   \n",
       "4                 0.004105             0.007120             0.011450   \n",
       "...                    ...                  ...                  ...   \n",
       "98825             0.007160             0.010200             0.014500   \n",
       "98826             0.004430             0.006190             0.008880   \n",
       "98827             0.004320             0.006340             0.009132   \n",
       "98828             0.004980             0.007060             0.010800   \n",
       "98829             0.004985             0.007360             0.009405   \n",
       "\n",
       "       dwelling_time_-1_mean  ...  dwelling_time_+1_75  dwelling_time_+1_mean  \\\n",
       "0                   0.006574  ...             0.007888               0.006340   \n",
       "1                   0.006866  ...             0.007065               0.005963   \n",
       "2                   0.008893  ...             0.009300               0.008132   \n",
       "3                   0.007647  ...             0.009065               0.007299   \n",
       "4                   0.008467  ...             0.008030               0.008201   \n",
       "...                      ...  ...                  ...                    ...   \n",
       "98825               0.011100  ...             0.008630               0.006633   \n",
       "98826               0.007500  ...             0.006270               0.005211   \n",
       "98827               0.007294  ...             0.008390               0.006698   \n",
       "98828               0.008229  ...             0.007300               0.006016   \n",
       "98829               0.008041  ...             0.007857               0.005726   \n",
       "\n",
       "       std_+1_25  std_+1_50  std_+1_75  std_+1_mean  mean_+1_25  mean_+1_50  \\\n",
       "0         1.5300      1.720     1.9825     1.804167      94.700        95.5   \n",
       "1         1.6100      2.105     2.5400     2.255000      81.800        82.2   \n",
       "2         2.4700      3.610     5.0800     4.088857      76.600        77.8   \n",
       "3         2.5100      3.460     4.6950     3.717714      79.900        80.9   \n",
       "4         1.8800      2.740     3.8750     2.998000      85.000        88.2   \n",
       "...          ...        ...        ...          ...         ...         ...   \n",
       "98825     2.0400      2.330     2.6900     2.416410      94.300        95.4   \n",
       "98826     2.3300      3.090     4.1400     3.472566      78.300        80.3   \n",
       "98827     2.0350      2.885     3.8300     3.127305      88.025        89.2   \n",
       "98828     2.0400      2.380     2.7700     2.483366      81.200        82.5   \n",
       "98829     1.5375      1.765     2.6825     2.047667      93.700        94.5   \n",
       "\n",
       "       mean_+1_75  mean_+1_mean  \n",
       "0          96.500     95.641667  \n",
       "1          83.375     82.560526  \n",
       "2          79.300     78.120000  \n",
       "3          84.050     81.951429  \n",
       "4          89.350     86.457143  \n",
       "...           ...           ...  \n",
       "98825      96.400     95.182051  \n",
       "98826      82.100     80.369912  \n",
       "98827      90.300     88.889831  \n",
       "98828      83.900     82.502970  \n",
       "98829      95.550     94.836667  \n",
       "\n",
       "[98830 rows x 42 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1_pre.feature_eng()\n",
    "df_2_pre.feature_eng()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>nucleotides</th>\n",
       "      <th>reads_count</th>\n",
       "      <th>label</th>\n",
       "      <th>dwelling_time_-1_25</th>\n",
       "      <th>dwelling_time_-1_50</th>\n",
       "      <th>dwelling_time_-1_75</th>\n",
       "      <th>dwelling_time_-1_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>dwelling_time_+1_75</th>\n",
       "      <th>dwelling_time_+1_mean</th>\n",
       "      <th>std_+1_25</th>\n",
       "      <th>std_+1_50</th>\n",
       "      <th>std_+1_75</th>\n",
       "      <th>std_+1_mean</th>\n",
       "      <th>mean_+1_25</th>\n",
       "      <th>mean_+1_50</th>\n",
       "      <th>mean_+1_75</th>\n",
       "      <th>mean_+1_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>155</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004235</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.006790</td>\n",
       "      <td>0.006574</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007888</td>\n",
       "      <td>0.006340</td>\n",
       "      <td>1.5300</td>\n",
       "      <td>1.720</td>\n",
       "      <td>1.9825</td>\n",
       "      <td>1.804167</td>\n",
       "      <td>94.700</td>\n",
       "      <td>95.5</td>\n",
       "      <td>96.500</td>\n",
       "      <td>95.641667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>165</td>\n",
       "      <td>ATAACCA</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004980</td>\n",
       "      <td>0.005705</td>\n",
       "      <td>0.008533</td>\n",
       "      <td>0.006866</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007065</td>\n",
       "      <td>0.005963</td>\n",
       "      <td>1.6100</td>\n",
       "      <td>2.105</td>\n",
       "      <td>2.5400</td>\n",
       "      <td>2.255000</td>\n",
       "      <td>81.800</td>\n",
       "      <td>82.2</td>\n",
       "      <td>83.375</td>\n",
       "      <td>82.560526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>347</td>\n",
       "      <td>AAGACAG</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.003835</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.010280</td>\n",
       "      <td>0.008893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009300</td>\n",
       "      <td>0.008132</td>\n",
       "      <td>2.4700</td>\n",
       "      <td>3.610</td>\n",
       "      <td>5.0800</td>\n",
       "      <td>4.088857</td>\n",
       "      <td>76.600</td>\n",
       "      <td>77.8</td>\n",
       "      <td>79.300</td>\n",
       "      <td>78.120000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>435</td>\n",
       "      <td>AAGACAA</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.006970</td>\n",
       "      <td>0.008465</td>\n",
       "      <td>0.007647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009065</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>2.5100</td>\n",
       "      <td>3.460</td>\n",
       "      <td>4.6950</td>\n",
       "      <td>3.717714</td>\n",
       "      <td>79.900</td>\n",
       "      <td>80.9</td>\n",
       "      <td>84.050</td>\n",
       "      <td>81.951429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>AT1G01050.1</td>\n",
       "      <td>549</td>\n",
       "      <td>ATGACAA</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004105</td>\n",
       "      <td>0.007120</td>\n",
       "      <td>0.011450</td>\n",
       "      <td>0.008467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008030</td>\n",
       "      <td>0.008201</td>\n",
       "      <td>1.8800</td>\n",
       "      <td>2.740</td>\n",
       "      <td>3.8750</td>\n",
       "      <td>2.998000</td>\n",
       "      <td>85.000</td>\n",
       "      <td>88.2</td>\n",
       "      <td>89.350</td>\n",
       "      <td>86.457143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98825</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>367</td>\n",
       "      <td>AAAACTC</td>\n",
       "      <td>117</td>\n",
       "      <td>2</td>\n",
       "      <td>0.007160</td>\n",
       "      <td>0.010200</td>\n",
       "      <td>0.014500</td>\n",
       "      <td>0.011100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008630</td>\n",
       "      <td>0.006633</td>\n",
       "      <td>2.0400</td>\n",
       "      <td>2.330</td>\n",
       "      <td>2.6900</td>\n",
       "      <td>2.416410</td>\n",
       "      <td>94.300</td>\n",
       "      <td>95.4</td>\n",
       "      <td>96.400</td>\n",
       "      <td>95.182051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98826</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>444</td>\n",
       "      <td>GTGACCC</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004430</td>\n",
       "      <td>0.006190</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006270</td>\n",
       "      <td>0.005211</td>\n",
       "      <td>2.3300</td>\n",
       "      <td>3.090</td>\n",
       "      <td>4.1400</td>\n",
       "      <td>3.472566</td>\n",
       "      <td>78.300</td>\n",
       "      <td>80.3</td>\n",
       "      <td>82.100</td>\n",
       "      <td>80.369912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98827</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>465</td>\n",
       "      <td>GTGACTC</td>\n",
       "      <td>118</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004320</td>\n",
       "      <td>0.006340</td>\n",
       "      <td>0.009132</td>\n",
       "      <td>0.007294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.006698</td>\n",
       "      <td>2.0350</td>\n",
       "      <td>2.885</td>\n",
       "      <td>3.8300</td>\n",
       "      <td>3.127305</td>\n",
       "      <td>88.025</td>\n",
       "      <td>89.2</td>\n",
       "      <td>90.300</td>\n",
       "      <td>88.889831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98828</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>663</td>\n",
       "      <td>GAAACCT</td>\n",
       "      <td>101</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004980</td>\n",
       "      <td>0.007060</td>\n",
       "      <td>0.010800</td>\n",
       "      <td>0.008229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.006016</td>\n",
       "      <td>2.0400</td>\n",
       "      <td>2.380</td>\n",
       "      <td>2.7700</td>\n",
       "      <td>2.483366</td>\n",
       "      <td>81.200</td>\n",
       "      <td>82.5</td>\n",
       "      <td>83.900</td>\n",
       "      <td>82.502970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98829</th>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67600.1</td>\n",
       "      <td>154</td>\n",
       "      <td>AGAACTC</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>0.004985</td>\n",
       "      <td>0.007360</td>\n",
       "      <td>0.009405</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007857</td>\n",
       "      <td>0.005726</td>\n",
       "      <td>1.5375</td>\n",
       "      <td>1.765</td>\n",
       "      <td>2.6825</td>\n",
       "      <td>2.047667</td>\n",
       "      <td>93.700</td>\n",
       "      <td>94.5</td>\n",
       "      <td>95.550</td>\n",
       "      <td>94.836667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98830 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id   transcript position nucleotides  reads_count  label  \\\n",
       "0            0  AT1G01050.1      155     GAAACTA           36      2   \n",
       "1            0  AT1G01050.1      165     ATAACCA           38      2   \n",
       "2            0  AT1G01050.1      347     AAGACAG           35      2   \n",
       "3            0  AT1G01050.1      435     AAGACAA           35      2   \n",
       "4            0  AT1G01050.1      549     ATGACAA           35      2   \n",
       "...        ...          ...      ...         ...          ...    ...   \n",
       "98825        0  AT5G67590.1      367     AAAACTC          117      2   \n",
       "98826        0  AT5G67590.1      444     GTGACCC          113      2   \n",
       "98827        0  AT5G67590.1      465     GTGACTC          118      2   \n",
       "98828        0  AT5G67590.1      663     GAAACCT          101      2   \n",
       "98829        0  AT5G67600.1      154     AGAACTC           30      2   \n",
       "\n",
       "       dwelling_time_-1_25  dwelling_time_-1_50  dwelling_time_-1_75  \\\n",
       "0                 0.004235             0.005310             0.006790   \n",
       "1                 0.004980             0.005705             0.008533   \n",
       "2                 0.003835             0.007300             0.010280   \n",
       "3                 0.005310             0.006970             0.008465   \n",
       "4                 0.004105             0.007120             0.011450   \n",
       "...                    ...                  ...                  ...   \n",
       "98825             0.007160             0.010200             0.014500   \n",
       "98826             0.004430             0.006190             0.008880   \n",
       "98827             0.004320             0.006340             0.009132   \n",
       "98828             0.004980             0.007060             0.010800   \n",
       "98829             0.004985             0.007360             0.009405   \n",
       "\n",
       "       dwelling_time_-1_mean  ...  dwelling_time_+1_75  dwelling_time_+1_mean  \\\n",
       "0                   0.006574  ...             0.007888               0.006340   \n",
       "1                   0.006866  ...             0.007065               0.005963   \n",
       "2                   0.008893  ...             0.009300               0.008132   \n",
       "3                   0.007647  ...             0.009065               0.007299   \n",
       "4                   0.008467  ...             0.008030               0.008201   \n",
       "...                      ...  ...                  ...                    ...   \n",
       "98825               0.011100  ...             0.008630               0.006633   \n",
       "98826               0.007500  ...             0.006270               0.005211   \n",
       "98827               0.007294  ...             0.008390               0.006698   \n",
       "98828               0.008229  ...             0.007300               0.006016   \n",
       "98829               0.008041  ...             0.007857               0.005726   \n",
       "\n",
       "       std_+1_25  std_+1_50  std_+1_75  std_+1_mean  mean_+1_25  mean_+1_50  \\\n",
       "0         1.5300      1.720     1.9825     1.804167      94.700        95.5   \n",
       "1         1.6100      2.105     2.5400     2.255000      81.800        82.2   \n",
       "2         2.4700      3.610     5.0800     4.088857      76.600        77.8   \n",
       "3         2.5100      3.460     4.6950     3.717714      79.900        80.9   \n",
       "4         1.8800      2.740     3.8750     2.998000      85.000        88.2   \n",
       "...          ...        ...        ...          ...         ...         ...   \n",
       "98825     2.0400      2.330     2.6900     2.416410      94.300        95.4   \n",
       "98826     2.3300      3.090     4.1400     3.472566      78.300        80.3   \n",
       "98827     2.0350      2.885     3.8300     3.127305      88.025        89.2   \n",
       "98828     2.0400      2.380     2.7700     2.483366      81.200        82.5   \n",
       "98829     1.5375      1.765     2.6825     2.047667      93.700        94.5   \n",
       "\n",
       "       mean_+1_75  mean_+1_mean  \n",
       "0          96.500     95.641667  \n",
       "1          83.375     82.560526  \n",
       "2          79.300     78.120000  \n",
       "3          84.050     81.951429  \n",
       "4          89.350     86.457143  \n",
       "...           ...           ...  \n",
       "98825      96.400     95.182051  \n",
       "98826      82.100     80.369912  \n",
       "98827      90.300     88.889831  \n",
       "98828      83.900     82.502970  \n",
       "98829      95.550     94.836667  \n",
       "\n",
       "[98830 rows x 42 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2_pre.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "40cd0d416b49876a6e9545be524abe39160420e585629e7d063418f197eadf97"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
